### PR TITLE
[Service] Fix for improving query parameter mapping in resource function

### LINKED ISSUE #424: Improve query parameter mapping in resource function
**Description:**
Ballerina language contains three types of main parameter types: required, defaultable, rest. We need to improve our query parameter mapping by aligning to the ballerina parameter. 

- [x] Issue with Nested array query parameter
- [x] OpenAPI has optional query parameter
- [x] OpenAPI has query parameter with default value
- [x] OpenAPI has query parameter with nullable true

**Suggested Labels:**

**Suggested Assignees:**

**Affected Product Version:**

**OS, DB, other environment details and versions:**    

**Steps to reproduce:**

**Related Issues:**

### PR DESCRIPTION
**Pull request by lnash94**
## Automation tests
 - Unit tests  - added
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: openapi-cli/src/main/java/io/ballerina/openapi/generators/service/ServiceDiagnosticMessages.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * This {@code ServiceDiagnosticMessages} enum class for containing the error message related to openapi to ballerina
+ * service generation command.
+ *
+ * @since 2.0.0
+ */
+public enum ServiceDiagnosticMessages {
+    OAS_SERVICE_100("OAS_SERVICE_100", "Ballerina resource functions are not support to query" +
+            " parameters with nested array type.", DiagnosticSeverity.ERROR),
```
Comment (NipunaRanasinghe) [inline openapi-cli/src/main/java/io/ballerina/openapi/generators/service/ServiceDiagnosticMessages.java:31]:
```suggestion
    OAS_SERVICE_100("OAS_SERVICE_100", "Query" +
            " parameters with nested array types are not supported in Ballerina.", DiagnosticSeverity.ERROR),
```

File: openapi-cli/src/main/java/io/ballerina/openapi/generators/service/ServiceDiagnosticMessages.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    OAS_SERVICE_101("OAS_SERVICE_101",
+            "Query parameter with no array item type can not be mapped to ballerina resource query parameter.",
```
Comment (NipunaRanasinghe) [inline openapi-cli/src/main/java/io/ballerina/openapi/generators/service/ServiceDiagnosticMessages.java:33]:
```suggestion
    OAS_SERVICE_101("OAS_SERVICE_101",
            "Query parameters with no array item type can not be mapped to Ballerina resource query parameters.",
```

File: openapi-cli/src/main/java/io/ballerina/openapi/generators/service/ServiceDiagnosticMessages.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            DiagnosticSeverity.ERROR),
+    OAS_SERVICE_102("OAS_SERVICE_102",
+            "Given OpenAPI query parameter type '%s' can not be mapped to the Ballerina query parameter.",
```
Comment (NipunaRanasinghe) [inline openapi-cli/src/main/java/io/ballerina/openapi/generators/service/ServiceDiagnosticMessages.java:36]:
```suggestion
            "query parameters with type '%s' can not be mapped to the Ballerina query parameters.",
```

File: openapi-cli/src/main/java/io/ballerina/openapi/generators/service/ServiceGenerationUtils.java [hunk 4]
```diff
@@ -222,4 +242,26 @@ public static TypeDescriptorNode getMediaTypeToken(Map.Entry<String, MediaType>
                 return createSimpleNameReferenceNode(identifierToken);
         }
     }
+
+    public static MetadataNode generateServiceConfigAnnotation() {
```
Comment (NipunaRanasinghe) [inline openapi-cli/src/main/java/io/ballerina/openapi/generators/service/ServiceGenerationUtils.java:246]: Lets add java docs as this is `public static` method

File: openapi-cli/src/main/java/io/ballerina/openapi/generators/service/ParametersGenerator.java [hunk 5]
```diff
@@ -157,46 +177,101 @@ private RequiredParameterNode handleHeader(Parameter parameter)throws BallerinaO
...
+            throw new BallerinaOpenApiException(String.format(messages.getDescription(), parameter.getName()));
+        } else if (parameter.getSchema().getDefault() != null) {
+            // When query parameter has default value
+            return handleDefaultQueryParameter(schema, annotations, parameterName);
+        } else if (parameter.getRequired() != null && parameter.getRequired() && schema.getNullable() == null) {
+            // Required typeDescriptor
+            return handleRequiredQueryParameter(schema, annotations, parameterName);
         } else {
+            // Optional typeDescriptor
+            return handleOptionalQueryParameter(schema, annotations, parameterName);
```
[the remark this answers is not in the document]
Comment (lnash94) [reply openapi-cli/src/main/java/io/ballerina/openapi/generators/service/ParametersGenerator.java:208]: In mainly, here consider the required parameter, defaultable parameter and rest behaviour map to an optional parameter , other special scenarios `nullable` will be handled inside the optional parameter.

**Review by NipunaRanasinghe — APPROVED** [review]

**Pull request merged**